### PR TITLE
feat: Phase 4.2 - Fix skip event path for flat mode in convertAllPosts

### LINKED ISSUE #52: 4.2 Update Converter convertAllPosts Method for Flat Mode
### Overview
Update the `convertAllPosts` method to properly handle flat mode by creating the shared image directory once and using the correct FileWriter for post existence checks.

### Tasks
- [x] Create FileWriter with correct output mode for `postExists` check
- [x] Create shared image directory once at start (for flat mode efficiency)
- [x] Pass `outputStructure` through to `convertPost`

### Proposed Changes

**File: `src/converter.ts`**

Update `convertAllPosts` method (around line 274):

```typescript
async convertAllPosts(
  exportPath: string,
  outputDir: string,
  options?: ConversionOptions
): Promise<ConversionResult> {
  // ... existing setup code ...

  const outputStructure = effectiveOptions.outputStructure ?? { mode: 'nested' };
  const isFlat = outputStructure.mode === 'flat';

  // Create FileWriter with correct mode for existence check
  const existenceChecker = isFlat
    ? new FileWriter({ outputMode: 'flat' })
    : this.fileWriter;

  // Create shared image directory once for flat mode (efficiency)
  if (isFlat) {
    const parentDir = path.dirname(outputDir);
    const imageFolderName = outputStructure.imageFolderName ?? '_images';
    const imageDir = path.join(parentDir, imageFolderName);
    if (!fs.existsSync(imageDir)) {
      fs.mkdirSync(imageDir, { recursive: true });
    }
  }

  // ... in the post loop:
  for (const post of posts) {
    const slug = this.extractSlugSafely(post, index);
    
    // Use correct existence checker based on mode
    if (effectiveOptions.skipExisting && existenceChecker.postExists(outputDir, slug)) {
      // ... existing skip logic
      continue;
    }

    // ... rest of loop
  }
}
```

### Key Changes

1. **Existence Check**: Uses flat-mode FileWriter to check for `{slug}.md` files instead of `{slug}/` directories
2. **Image Directory**: Created once at start to avoid repeated checks per post
3. **Backwards Compatible**: Defaults to nested mode if `outputStructure` not specified

### References
- Implementation Plan: [docs/IMPLEMENTATION_FLAT.md](../docs/IMPLEMENTATION_FLAT.md)
- Depends on: #51 (convertPost updates)

### PR DESCRIPTION
**Pull request by alvincrespo**
## Phase 4.2: Update Converter convertAllPosts Method for Flat Mode

Fixes #52

## Overview

This PR completes Phase 4.2 of the flat output mode implementation. After thorough code analysis, most of the originally planned changes were already complete from Phase 4.1. This PR fixes the only remaining bug: skip events emitting hardcoded nested paths regardless of output mode.

## Bug Fixed 🐛

**Issue**: Skip event `outputPath` was hardcoded to nested format (`{slug}/index.md`) on line 335, regardless of the actual output mode.

**Solution**: Created `getSkipOutputPath()` helper method that leverages the existing `Post.getFilePath()` logic for mode-aware path resolution.

### Core Implementation
- ✅ Added `Post` model import to [src/converter.ts](src/converter.ts)
- ✅ Created `getSkipOutputPath()` helper method (lines 860-875)
- ✅ Updated skip event creation to use helper method (line 336)
- ✅ Reuses existing, tested path logic from Post model (no duplication)

### Test Coverage
Added 3 comprehensive integration tests ([tests/integration/converter.test.ts](tests/integration/converter.test.ts)):

1. **Skip event path - nested mode**: Validates backward compatibility
2. **Skip event path - flat mode**: Validates flat mode behavior
3. **Invalid slug edge case**: Validates error handling with fallback to nested format

### Documentation
- ✅ Updated [docs/IMPLEMENTATION_FLAT.md](docs/IMPLEMENTATION_FLAT.md) Step 4.2 status
- ✅ Created detailed implementation plan at [docs/features/flat/PHASE_4_STEP_4_2.md](docs/features/flat/PHASE_4_STEP_4_2.md)

## Verification ✅

```bash
# TypeScript compilation
npm run type-check
# ✅ No errors

# Test suite
npm test
# ✅ All 450 tests passing (was 447, added 3 new tests)
# ✅ Coverage: 99.49% (exceeds 99% target)

# Build
npm run build
# ✅ Successful
```

## Key Discoveries

During implementation, I found that:
1. ✅ `FileWriter` is already mode-aware (created at construction with correct `outputMode`)
2. ✅ `postExists()` is already mode-aware (uses `Post.getExistencePath()` internally)
3. ✅ Image directories are created correctly (per-post in `convertPostFlat`)

**The only actual issue was the hardcoded skip path**, which is now fixed.

### Before (Bug)
```typescript
// Skip event always used nested path
{
  outputPath: path.join(outputDir, slug, 'index.md')  // ❌ Always nested
}
```

### After (Fixed)
```typescript
// Skip event uses mode-aware path
// Nested mode: /output/slug/index.md
// Flat mode:   /output/slug.md
{
  outputPath: this.getSkipOutputPath(outputDir, slug)  // ✅ Mode-aware
}
```

## Backward Compatibility

- ✅ Nested mode behavior unchanged (still uses `{slug}/index.md`)
- ✅ All existing tests pass without modification
- ✅ Default behavior remains nested mode

## Next Steps

After merging this PR:
- [ ] Continue to Phase 4.3: Write Converter Integration Tests (if needed)
- [ ] Continue to Phase 5: CLI Updates
- [ ] Update GitHub issue #52 status

## Testing Instructions

```bash
# Run tests
npm test

# Test nested mode (default)
npx ts-node src/cli/convert.ts --export sample.json --output ./blog

# Test flat mode
npx ts-node src/cli/convert.ts --export sample.json --output ./blog --flat
```

---

**Ready for review!** This is a focused bug fix with comprehensive test coverage and clear documentation.

### CODE REVIEW

File: tests/integration/converter.test.ts [hunk 1]
```diff
@@ -283,6 +283,53 @@ describe('Converter', () => {
...
+
+      // Should fall back to nested format for invalid slugs
+      expect(completedHandler).toHaveBeenCalledWith(
+        expect.objectContaining({
+          result: expect.objectContaining({
+            outputPath: expect.stringMatching(/index\.md$/),
+          }),
+        })
+      );
+    });
```
[the remark this answers is not in the document]
Comment (alvincrespo) [reply tests/integration/converter.test.ts]: Addressed in ac49f29

File: docs/features/flat/PHASE_4_STEP_4_2.md [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+- [x] Run `npm run type-check` - verify no errors
+- [x] Run `npm run build` - verify success
+- [x] Run `npm test` - verify all tests pass
+- [x] Run `npm run test:coverage` - verify ≥99% coverage
+
+### Phase 4: Documentation
+- [x] Mark Step 4.2 complete in IMPLEMENTATION_FLAT.md
+- [x] Update GitHub issue #52 with completion status
+- [x] Create this plan document at docs/features/flat/PHASE_4_STEP_4_2.md
+
```
[the remark this answers is not in the document]
Comment (alvincrespo) [reply docs/features/flat/PHASE_4_STEP_4_2.md:376]: Addressed by cd2bb07

File: tests/integration/converter.test.ts [hunk 3]
```diff
@@ -283,6 +284,64 @@ describe('Converter', () => {
...
+      expect(completedHandler).toHaveBeenCalledWith(
+        expect.objectContaining({
+          result: expect.objectContaining({
+            slug: 'test-post',
+            outputPath: path.join('/output', 'test-post', 'index.md'),
+            success: true,
+          }),
+        })
+      );
+    });
```
Comment (alvincrespo) [inline tests/integration/converter.test.ts:306]: Good catch on the documentation reference! The plan did originally call for that test, but we evolved the approach during implementation. Here's what happened:

## Original Plan vs. Final Implementation

**Original plan**: Test "invalid slugs in skip path gracefully" with an invalid slug scenario

**Issue identified**: That scenario is unrealistic because:
- `FileWriter.postExists()` internally creates a `Post` instance (file-writer.ts:228-234)
- Invalid slugs throw `PostValidationError` which is caught (line 237-238)
- `postExists()` returns `false` for invalid slugs
- The skip logic never executes for invalid slugs

**Decision**: Removed the unrealistic test and replaced it with proper mocking

## Current Implementation (Better)

**Test added** (lines 307-338): "should fall back to nested format when Post.getFilePath fails"

This test:
- Uses proper mocking (mocks `Post.getFilePath()` to throw an error)
- Tests the defensive error handling in `getSkipOutputPath()`
- Verifies warning is logged
- Verifies fallback to nested format works
- Documents it's testing defensive programming through mocking

This approach is more honest - it uses mocking to trigger the failure path rather than creating an impossible scenario that couldn't occur in production.

The documentation reference is to the original plan; the actual implementation improved on it.

File: tests/integration/converter.test.ts [hunk 4]
```diff
@@ -283,6 +283,26 @@ describe('Converter', () => {
...
```
[the remark this answers is not in the document]
Comment (alvincrespo) [reply tests/integration/converter.test.ts:306]: reference: https://github.com/alvincrespo/hashnode-content-converter/pull/85#discussion_r2771555116
